### ISSUE TITLE
RegExpObj.test() with regex pattern '\1' failed in nginx, but it's execute normal in njs-cli

### ISSUE THREAD

**Issue by otherNoscript** [nginx/njs#225]
Hello. I try run simple function in nginx docker images with latest and alpine tags

**nginx.conf**
```
load_module modules/ngx_http_js_module-debug.so;

user              nginx;
worker_processes  auto;

error_log  /var/log/nginx/error.log debug;
pid        /var/run/nginx.pid;


events {
    worker_connections  1024;
}

http {
    include            /etc/nginx/mime.types;
    #default_type      application/octet-stream;

    log_format  main   '$remote_addr - $remote_user [$time_local] "$request" '
                       '$status $body_bytes_sent "$http_referer" '
                       '"$http_user_agent" "$http_x_forwarded_for"';

    access_log  /var/log/nginx/access.log  main;

    sendfile            on;
    #tcp_nopush         on;

    gzip                on;
    gzip_disable        "MSIE [1-6]\\.(?!.*SV1)";
    gzip_http_version   1.1;
    gzip_proxied        any;
    gzip_comp_level     5;
    gzip_types          text/plain    text/css application/javascript application/x-javascript text/xml application/xml application/rss+xml text/javascript image/x-icon image/bmp image/svg+xml;
    gzip_vary           on;
    #keepalive_timeout  65;

    js_include          http.js;

    server {
        listen 8080;
        server_name _;

        location / {
            js_content dbg_req;
        }
    }
}
```

**http.js**
```js
// 1. var re = /<(?<key>[\w\-\.\:]+)\s*(?<tags>[^>]*)>(?<body>.*?)<\/[^>]*>/g;
// 2. var re = /<(?<key>[\w\-\.\:]+)\s*(?<tags>[^>]*)>(?<body>.*?)<\/\1>/g;

function dbg_req(r) {
    var data = '<Name>Text</Name>',
        reTest = new RegExp(re);

    r.return(200, reTest.test(data));
}
```

If i use first pattern function return status_code `200` and body `true`.
If i use second pattern function return status_code `500`.
But if i run `reTest.test(data)` in the njs-cli in the both cases the correct value will be returned.

RegExpObj.exec() with second pattern work normally in nginx.

**Comment by xeioex** [comment]
Hi @otherNoscript,

Thanks for reporting it.
I can confirm that the issue is real, it will be fixed. It is only related to `Regexp.prototype.test()` with regexp containing backreferences (`\1` for example). As a workaround  `Regexp.prototype.exec()` can be used.

**Comment by xeioex** [comment]
@otherNoscript welcome to test

```diff
# HG changeset patch
# User Dmitry Volyntsev <xeioex@nginx.com>
# Date 1570122060 -10800
#      Thu Oct 03 20:01:00 2019 +0300
# Node ID 0a7043157035c0c97658fd3a016eb468c06805b4
# Parent  956b24477d59ec86a2faa3e65a92f95c2b6fda0b
Fixed Regexp.prototype.test() for regexps with backreferences.

This closes #225 issue on Github.

diff --git a/src/njs_pcre.c b/src/njs_pcre.c
--- a/src/njs_pcre.c
+++ b/src/njs_pcre.c
@@ -111,6 +111,17 @@ njs_regex_compile(njs_regex_t *regex, u_
         goto done;
     }
 
+    err = pcre_fullinfo(regex->code, NULL, PCRE_INFO_BACKREFMAX,
+                        &regex->backrefmax);
+
+    if (njs_slow_path(err < 0)) {
+        njs_alert(ctx->trace, NJS_LEVEL_ERROR,
+                  "pcre_fullinfo(\"%s\", PCRE_INFO_BACKREFMAX) failed: %d",
+                  pattern, err);
+
+        goto done;
+    }
+
     /* Reserve additional elements for the first "$0" capture. */
     regex->ncaptures++;
 
@@ -175,6 +186,13 @@ njs_regex_ncaptures(njs_regex_t *regex)
 }
 
 
+njs_uint_t
+njs_regex_backrefmax(njs_regex_t *regex)
+{
+    return regex->backrefmax;
+}
+
+
 njs_int_t
 njs_regex_named_captures(njs_regex_t *regex, njs_str_t *name, int n)
 {
diff --git a/src/njs_pcre.h b/src/njs_pcre.h
--- a/src/njs_pcre.h
+++ b/src/njs_pcre.h
@@ -18,6 +18,7 @@ struct njs_regex_s {
     pcre        *code;
     pcre_extra  *extra;
     int         ncaptures;
+    int         backrefmax;
     int         nentries;
     int         entry_size;
     char        *entries;
diff --git a/src/njs_regex.h b/src/njs_regex.h
--- a/src/njs_regex.h
+++ b/src/njs_regex.h
@@ -31,6 +31,7 @@ NJS_EXPORT njs_int_t njs_regex_compile(n
     size_t len, njs_uint_t options, njs_regex_context_t *ctx);
 NJS_EXPORT njs_bool_t njs_regex_is_valid(njs_regex_t *regex);
 NJS_EXPORT njs_uint_t njs_regex_ncaptures(njs_regex_t *regex);
+NJS_EXPORT njs_uint_t njs_regex_backrefmax(njs_regex_t *regex);
 NJS_EXPORT njs_int_t njs_regex_named_captures(njs_regex_t *regex,
     njs_str_t *name, int n);
 NJS_EXPORT njs_regex_match_data_t *njs_regex_match_data(njs_regex_t *regex,
diff --git a/src/njs_regexp.c b/src/njs_regexp.c
--- a/src/njs_regexp.c
+++ b/src/njs_regexp.c
@@ -842,11 +842,13 @@ static njs_int_t
 njs_regexp_prototype_test(njs_vm_t *vm, njs_value_t *args, njs_uint_t nargs,
     njs_index_t unused)
 {
-    njs_int_t             ret;
-    njs_uint_t            n;
-    const njs_value_t     *value, *retval;
-    njs_string_prop_t     string;
-    njs_regexp_pattern_t  *pattern;
+    njs_int_t               ret;
+    njs_uint_t              n;
+    njs_regex_t             *regex;
+    const njs_value_t       *value, *retval;
+    njs_string_prop_t       string;
+    njs_regexp_pattern_t    *pattern;
+    njs_regex_match_data_t  *match_data;
 
     if (!njs_is_regexp(njs_arg(args, nargs, 0))) {
         njs_type_error(vm, "\"this\" argument is not a regexp");
@@ -866,20 +868,42 @@ njs_regexp_prototype_test(njs_vm_t *vm, 
 
     pattern = njs_regexp_pattern(&args[0]);
 
-    if (njs_regex_is_valid(&pattern->regex[n])) {
-        ret = njs_regexp_match(vm, &pattern->regex[n], string.start,
-                               string.size, vm->single_match_data);
+    regex = &pattern->regex[n];
+
+    if (njs_regex_is_valid(regex)) {
+        if (njs_regex_backrefmax(regex)) {
+            match_data = njs_regex_match_data(regex, vm->regex_context);
+            if (njs_slow_path(match_data == NULL)) {
+                njs_memory_error(vm);
+                return NJS_ERROR;
+            }
+
+        } else {
+            match_data = vm->single_match_data;
+        }
+
+        ret = njs_regexp_match(vm, regex, string.start, string.size,
+                               match_data);
         if (ret >= 0) {
             retval = &njs_value_true;
 
         } else if (ret != NJS_REGEX_NOMATCH) {
-            return NJS_ERROR;
+            ret = NJS_ERROR;
+            goto done;
         }
     }
 
+    ret = NJS_OK;
+
     vm->retval = *retval;
 
-    return NJS_OK;
+done:
+
+    if (match_data != vm->single_match_data) {
+        njs_regex_match_data_free(match_data, vm->regex_context);
+    }
+
+    return ret;
 }
 
 
diff --git a/src/test/njs_unit_test.c b/src/test/njs_unit_test.c
--- a/src/test/njs_unit_test.c
+++ b/src/test/njs_unit_test.c
@@ -8178,6 +8178,10 @@ static njs_unit_test_t  njs_test[] =
       njs_str("true") },
 #endif
 
+    { njs_str("var re = /<(?<key>[\\w\\-\\.\\:]+)>(?<body>.*?)<\\/\\1>/g;"
+              "['<A>XXX</A>', '<A>XX</B>'].map(s=>re.test(s))"),
+      njs_str("true,false") },
+
     { njs_str("/\\x80/.test('\\u0080')"),
       njs_str("true") },
 
```
